### PR TITLE
mem-ruby, configs: Add a CHI-TLM controller + testing

### PR DESCRIPTION
**Pull request by giactra**
This PR is implement a CacheController which acts as a bridge between the
AMBA TLM 2.0 implementation of CHI [[1]](https://developer.arm.com/documentation/101459/latest) [[2](https://developer.arm.com/Architectures/AMBA#Downloads)] with the gem5 (ruby) one.

In other words it translates AMBA CHI transactions into ruby
messages (which are then forwarded to the MessageQueues)
and viceversa.

```
ARM::CHI::Payload,         CHIRequestMsg
                     <-->  CHIDataMsg
ARM::CHI::Phase            CHIResponseMsg
                           CHIDataMsg
```

To be able to do so, it first adds a CHIGenericController class to model a coherence controller entirely
in C++ to work with ruby. Adding a CHIGenericController removes the limitation of using the gem5 Sequencer
and associated data structures, therefore allowing the aforementioned CHI to CHI bridge/controller
and potentially a hybrid setup with classic caches + ruby.

The PR also adds a CHI tester, which is a transaction generator directly connected to the CHI-TLM controller. It will
be possible to write small unit tests in python (the library has been wrapped with pybind) for CHI transactions and unlock unit-testing for ruby

### CODE REVIEW

**Comment by powerjg** [comment]
Would you be willing to talk about this in the next gem5-dev meeting?

This is super cool!

**Comment by giactra** [comment]
> Would you be willing to talk about this in the next gem5-dev meeting?

Will it be on the 9th of May? If that's the case I won't be able to attend it, though @tiagormk (who's the author of these changes/feature) might be

**Comment by ivanaamit** [comment]
Yes, it will be on May 9th at 4pm UTC. 

@tiagormk, would you be available to attend and discuss this improvement?

**Comment by tiagormk** [comment]
> Yes, it will be on May 9th at 4pm UTC.
> 
> @tiagormk, would you be available to attend and discuss this improvement?

Unfortunately I have a conflict that time. But I could join the June meeting probably.

**Comment by shinezyy** [comment]
Cool!
We are doing something similar: see gem5 issue #1164.

I notice that you have many empty implementations like 
```
Sequencer*
CHIGenericController::getCPUSequencer() const
{
    // CHIGenericController doesn't have a CPUSequencer
    return nullptr;
}
```

We are facing similar things. 
The root cause is that  `AbstractController` is the interface for a Ruby cache but not the interface for `an upstream` of Ruby cache. So we define a more basic class HasDownStream (has ruby cache downstream)  [here](https://github.com/shinezyy/gem5-1/tree/Extract-HasDownStream). This is only a draft.

I think there should be something defining a cleaner interface of `an upstream` of Ruby cache. I'd like to participate.

**Comment by giactra** [comment]
> Cool! We are doing something similar: see gem5 issue #1164.
> 

Yes I saw the issue few days ago and I also noticed the similarity

> I notice that you have many empty implementations like
> 
> ```
> Sequencer*
> CHIGenericController::getCPUSequencer() const
> {
>     // CHIGenericController doesn't have a CPUSequencer
>     return nullptr;
> }
> ```
> 

Yes, that is mainly because this controller constitutes a pure virtual interface. We have a subclass we will upstream which overrides most of these methods to be able to transform a gem5 classic packet into a ruby packet. The reason why we don't commit to a sequencer in the CHIGenericController is because we want to be able to support the case of a sequencer-less configuration.

If I understood correctly this might be a useful scenario for you as well.

> We are facing similar things. The root cause is that `AbstractController` is the interface for a Ruby cache but not the interface for `an upstream` of Ruby cache. So we define a more basic class HasDownStream (has ruby cache downstream) [here](https://github.com/shinezyy/gem5-1/tree/Extract-HasDownStream). This is only a draft.
> 
> I think there should be something defining a cleaner interface of `an upstream` of Ruby cache. I'd like to participate.

Thanks I will look at your draft if I find some time. I will also keep you in the loop if there is an update. Most likely we will discuss this feature at the next gem5 developer meeting.

**Comment by ivanaamit** [comment]
@Giactra, @Tiagormk: Are you available to present this at next week's developer meeting on Thursday, June 13th, at 4 PM UTC?

**Comment by tiagormk** [comment]
> @giactra, @tiagormk: Are you available to present this at next week's developer meeting on Thursday, June 13th, at 4 PM UTC?

Yes, I can present this.

**Comment by BobbyRBruce** [comment]
Is this waiting for review or are we still working on this?

**Comment by giactra** [comment]
> Is this waiting for review or are we still working on this?

Hi Bobby,
We discussed this few developers meetings ago; the PR is actually ready for review, though @powerjg was asking to extend the PR with a use case for this CHIGenericController. We will post this anytime soon, but if you want to start reviewing this I am definitely not gonna stop you ;)

**Comment by BobbyRBruce** [comment]
Next week (Wednesday) will be creating the Staging branch. Is this PR something you can complete soon and wish to get in?

I'm still a little unsure what needs done here but if I can help let just know

**Comment by giactra** [comment]
> Next week (Wednesday) will be creating the Staging branch. Is this PR something you can complete soon and wish to get in?
> 
> I'm still a little unsure what needs done here but if I can help let just know

Thanks Bobby,

As I mentioned in my prior comment, I have posted an update to the PR with the requested usecase

> We discussed this few developers meetings ago; the PR is actually ready for review, though @powerjg was asking to extend the PR with a use case for this CHIGenericController. We will post this anytime soon, but if you want to start reviewing this I am definitely not gonna stop you ;)

As the usecase is quite significant (in terms of codesize) I had to modify the PR title and message to reflect the new scope of the change. I would have preferred to do this in a separate PR and get this merged long time ago but the upstream consensus was to have definition and usage in one big PR... I apologise for the confusion this might cause.

Anyway, the PR is now ready for review

File: src/mem/ruby/protocol/chi/tlm/controller.cc
Comment (tiagormk) [inline src/mem/ruby/protocol/chi/tlm/controller.cc:320]: This (and the other `send*Msg` functions) currently assumes the underlying MessageBuffer is configured with unlimited capacity. I don't think this is a blocker for this PR, but it would be good to have a followup PR that adds some sort of link-level credit management to properly handle backpressure.

For instance, in CHIGenericController we could have a `virtual bool recvLinkCrd(unsigned crds, CHIChannel channel)` (or a separate one per channel type) that top-level can optionally implement to receive credits. It could be called from `CHIGenericController::init or initNetQueues` to return the initial number of credits (or 0 or UINT_MAX if the buffer is infinite), and every time a message is dequeued. We would need to register a callback with `MessageBuffer::registerDequeueCallback` to be informed of a dequeue.

I can create an issue to track this once it is merged.

**Review by tiagormk — APPROVED** [review]
Thanks @giactra !

LGTM (not sure it it counts as I've authored some of the commits). Also left a comment on better handling of backpressure, which is something that I recommend addressing in a future PR.

**Comment by giactra** [comment]
Hi @powerjg @BobbyRBruce , can I have an update on whether there is a chance this will be merged in 24.1?

**Comment by BobbyRBruce** [comment]
> Hi @powerjg @BobbyRBruce , can I have an update on whether there is a chance this will be merged in 24.1?

Yes! I've logged this as a PR we're going to let in directly to staging. I'm going to make a post about it soon, but essentially there are about 5 or 6 PRs we really want in but need work (or just reviews or testing), so while staging now exists (I've yet to make an announcement, but you can see it's a branch currently in the repo), we're going to permit some PRs (this one included)  to be merged into staging and therefore make it into the release.

Just seems like this release we have a lot of really nice features that just need just a _little more_ time, so we're going to give that time to PRs like this. WE'd really like to get most of these in over the next couple of weeks ideally though.

**Comment by giactra** [comment]
> > Hi @powerjg @BobbyRBruce , can I have an update on whether there is a chance this will be merged in 24.1?
> 
> Yes! I've logged this as a PR we're going to let in directly to staging. I'm going to make a post about it soon, but essentially there are about 5 or 6 PRs we really want in but need work (or just reviews or testing), so while staging now exists (I've yet to make an announcement, but you can see it's a branch currently in the repo), we're going to permit some PRs (this one included) to be merged into staging and therefore make it into the release.
> 
> Just seems like this release we have a lot of really nice features that just need just a _little more_ time, so we're going to give that time to PRs like this. WE'd really like to get most of these in over the next couple of weeks ideally though.

Thanks Bobby, makes sense. Let me know if I can help somehow if there's anything that needs reviewing

**Comment by giactra** [comment]
Hi @BobbyRBruce and @powerjg please have a look at the new iteration of the PR.
I have basically added some documentation on the Controller class and I stopped using Ruby.create_system as requested.

Please note: I am still using Ruby.setup_memory_controller. Let me know if you want me to remove that usage as well

**Review by BobbyRBruce — APPROVED** [review]

**Pull request merged**
